pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -235,6 +236,32 @@ func (r Results) ReservedOfferingErrors() map[*corev1.Pod]error {
 	})
 }
 
+func (r Results) NodePoolToPodMapping() map[string][]*corev1.Pod {
+	result := make(map[string][]*corev1.Pod)
+
+	for k, v := range lo.SliceToMap(r.NewNodeClaims, func(n *NodeClaim) (string, []*corev1.Pod) {
+		return n.Labels[v1.NodePoolLabelKey], n.Pods
+	}) {
+		result[k] = v
+	}
+
+	for k, v := range lo.SliceToMap(lo.Filter(r.ExistingNodes, func(n *ExistingNode, _ int) bool {
+		// Filter out nodes that don't have the nodePool label
+		return n.Labels()[v1.NodePoolLabelKey] != ""
+	}), func(n *ExistingNode) (string, []*corev1.Pod) {
+		return n.Labels()[v1.NodePoolLabelKey], n.Pods
+	}) {
+		if existing, ok := result[k]; ok {
+			// If key exists, append the values
+			result[k] = append(slices.Clone(existing), v...)
+		} else {
+			// If key doesn't exist, add the new key-value pair
+			result[k] = v
```

Comment:
`slices.Clone()` here you don't need to do it on L256 if you just clone it at the beginning